File: examples/Phase4Test.cs
using System;
using Roblox;
using Roblox.Instances;

namespace Examples;

[RobloxStruct]
public struct Point
{
    public float X { get; set; }
    public float Y { get; set; }
}

public struct Color
{
    public float R, G, B;
}

public class Phase4Test
{
    public void Run()
    {
        // Test Struct Emission & Initializers
        var p1 = new Point { X = 10, Y = 20 };
        
        // Test Clone Inserter (Assignment)
        var p2 = p1; 
        p2.X = 100; // p1.X should still be 10 in C#, so p2 must be a clone in Luau

        // Test Clone Inserter (Arguments)
        PrintPoint(p1);

        // Test With Expression
        var p3 = p1 with { Y = 50 };

        // Test Pattern Matching (is)
        var part = new Part();
        if (part is BasePart bp)
        {
            Console.WriteLine("It is a BasePart: " + bp.Name);
        }

        // Test Pattern Matching (switch)
        object obj = part;
        switch (obj)
        {
            case BasePart p:
                Console.WriteLine("Switch: BasePart " + p.Name);
                break;
            case string s:
                Console.WriteLine("Switch: string " + s);
                break;
            default:
                Console.WriteLine("Switch: default");
                break;
        }
    }

    private void PrintPoint(Point p)
    {
        Console.WriteLine($"Point: {p.X}, {p.Y}");
    }
}

File: examples/Phase4Test.lua
local Point = {}
Point.__index = Point
function Point:Clone()
    return setmetatable(table.clone(self), Point)
end
function Point.new()
    local self = setmetatable({}, Point)
    return self
end
local Color = {}
Color.__index = Color
function Color:Clone()
    return setmetatable(table.clone(self), Color)
end
function Color.new()
    local self = setmetatable({}, Color)
    return self
end
local Phase4Test = {}
Phase4Test.__index = Phase4Test
function Phase4Test.new()
    local self = setmetatable({}, Phase4Test)
    return self
end
function Phase4Test:Run()
    local p1 = Point.new()
    local p2 = p1:Clone()
    p2.X = 100
    self:PrintPoint(p1:Clone())
    local p3 = function(_c)
        _c.Y = 50
        return _c
    end(p1:Clone())
    local part = Instance.new("Part")
    if part:IsA("BasePart") then
        local bp = part
        print("It is a BasePart: " .. bp.Name)
    end
    local obj = part
    if obj:IsA("BasePart") then
        local p = obj
        print("Switch: BasePart " .. p.Name)
    else
        print("Switch: default")
    end
end
function Phase4Test:PrintPoint(p)
    print(string.format("Point: %s, %s", p.X, p.Y))
end
return Phase4Test
